src/cpp/fastrtps_deprecated/publisher/PublisherHistory.cpp

```diff
@@ -478,5 +477,33 @@ bool PublisherHistory::is_key_registered(
            );
 }
 
+bool PublisherHistory::wait_for_acknowledgement_last_change(
+        const InstanceHandle_t& handle,
+        const Duration_t& max_wait)
+{
+    if (WITH_KEY == topic_att_.getTopicKind())
+    {
+        t_m_Inst_Caches::iterator vit;
+        // Find the instance
+        if (find_or_add_key(handle, &vit))
```

Comment:
This is ensure because `DataWriterImpl::wait_for_acknowledgments` calls `PublisherHistory::is_key_registered` and if it is not, the method returns without calling `PublisherHistory::find_or_add_key`. Consequently, this method will always find (and not add).

On the other hand, if it is confusing I can just copy the find code of the method here or refactor this method so it has the find and the add part separated.

Reply:
@MiguelCompany, what do you think?

Reply:
Then we should just find the handle in `keyed_changes_` and check it is present. See my suggestion.